DependencyInjectionContainer/DIContainer.cs

```diff
@@ -1,130 +1,162 @@
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
+﻿using System.Reflection;
+using System.Data;
+using DependencyInjectionContainer.Exceptions;
 
 namespace DependencyInjectionContainer
 {
-    public class DIContainer
+    public sealed record DIContainer
     {
-        private List<Service> services;
+        private DIContainer containerParent;
 
-        public DIContainer()
+        internal DIContainer(List<Service> _services, DIContainer _containerParent)
         {
-            services = new List<Service>();
+            Services = _services;
+            containerParent = _containerParent;
         }
 
-        public void Register<TIType, TImplementation>(ServiceLifetime lifetime) where TImplementation : TIType
+        internal List<Service> Services { get; private set; }
+
+        public DIContainerBuilder CreateAChildContainer()
         {
-            ThrowExceptionIfServiceWithTypeExists(typeof(TImplementation));
-            services.Add(new Service(typeof(TImplementation), lifetime, typeof(TIType)));
+            return new DIContainerBuilder(this);
         }
 
-        public void Register<TImplementation>(ServiceLifetime lifetime)
+        public TypeToResolve Resolve<TypeToResolve>(ResolveSource resolveType = ResolveSource.Any) where TypeToResolve : class?
         {
-            if(typeof(TImplementation).IsAbstract)
+            if (IsEnumerable(typeof(TypeToResolve)))
             {
-                throw new ArgumentException("Can't register type without assigned implementation type");
+                return (TypeToResolve)InvokeResolveManyForIEnumerable(typeof(TypeToResolve), this, resolveType);
             }
-            ThrowExceptionIfServiceWithTypeExists(typeof(TImplementation));
-            services.Add(new Service(typeof(TImplementation), lifetime));
-        }
 
-        public void Register(object implementation, ServiceLifetime lifetime)
-        {
-            ThrowExceptionIfServiceWithTypeExists(implementation.GetType());
-            services.Add(new Service(implementation, lifetime));
-        }
+            if (resolveType == ResolveSource.NonLocal)
+            {
+                return ResolveNonLocal();
+            }
 
+            var servicesToResolve = Services.Where(service => IsServiceOfGivenType(service, typeof(TypeToResolve)));
 
-        private object Resolve(Type typeToResolve)
-            //if TypeToResolve is abstract resolves the first object which implements TypeToResolve
-        {
-            Service serviceToResolve = services.FirstOrDefault(service => IsServiceOfGivenType(service, typeToResolve));
+            if (servicesToResolve.Count() > 1)
+            {
+                throw new ArgumentException($"Many services with type {typeof(TypeToResolve)} was registered. Use ResolveMany to resolve them all");
+            }
 
-            if (serviceToResolve == null)
+            if (servicesToResolve.Count() == 0)
```

Comment:
Here will be multiple enumeration

https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1851